### ISSUE TITLE
Outdated documentation for Compilation

### ISSUE THREAD

**Issue by ivanskodje** [kwhat/jnativehook#327]
Greetings!

TLDR; **Could the Compiling.md documentation please be updated so I (and other new devs) can get started with the project?** :) 

I am attempting to compile and build the project, however I am incredibly new working with C++ type applications, and have trouble compiling when following the set instructions from /doc/Compiling.md.

I have installed (or so I believe) the required C++ tools, and so on (GCC, CMAKE, LLVM, MSVM, etc.) and is able to run the CMakeList.txt. However, for that to even work when following the Compiling.md instructions, I had to move it from root, to  ...\external\libuiohook\.

However, I am able to run the normal `mvn compile`, but not the second command:

```cmake
cmake -B target/build/libuiohook/ -S src/external/libuiohook/ -D CMAKE_INSTALL_PREFIX=target/ -D CMAKE_VERBOSE_MAKEFILE=ON -D BUILD_SHARED_LIBS=OFF
```

I am getting the following error response:

> -- Building for: Visual Studio 16 2019
-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19042.
-- The C compiler identification is MSVC 19.28.29912.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29910/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
CMake Error at CMakeLists.txt:47 (message):
  Could not determine target architecture.  Please set CMAKE_SYSTEM_PROCESSOR
  to the desired resource folder.
 
 I have attempted to add the missing variable by adding `-D CMAKE_SYSTEM_PROCESSOR=x86_64`, but it does not seem to work either way. 
You can find the [CMakeOutput.log here](https://pastebin.com/1crQXCQ0).
 
**Could the Compiling.md documentation please be updated?**

Thank you for understanding.

**Comment by kwhat** [comment]
Docs should be updated, although its pretty short right now.  I have copied the windows x86_64 compiling commands I am using with github actions below. 

        mvn compile
        
        md target\build\libuiohook\
        
        cmake -B target\build\libuiohook\  -S src\external\libuiohook\ ^
            -G "Visual Studio 16 2019" -A x64 ^
            -D CMAKE_INSTALL_PREFIX=target\ ^
            -D CMAKE_VERBOSE_MAKEFILE=ON ^
            -D BUILD_SHARED_LIBS=OFF
        
        cmake --build target\build\libuiohook\ ^
            --parallel 2 ^
            --config RelWithDebInfo ^
            --clean-first
        
        cmake --install target\build\libuiohook\ ^
            --config RelWithDebInfo
        
        md target\build\jni\
        
        cmake -B target\build\jni\ ^
            -G "Visual Studio 16 2019" -A x64 ^
            -D CMAKE_PREFIX_PATH=target\ ^
            -D CMAKE_SYSTEM_NAME=windows ^
            -D CMAKE_SYSTEM_PROCESSOR=x86_64 ^
            -D CMAKE_VERBOSE_MAKEFILE=ON
        
        cmake --build target\build\jni\ ^
            --parallel 2 ^
            --config RelWithDebInfo ^
            --clean-first
        
        cmake --install target\build\jni\ ^
            --config RelWithDebInfo

        mvn package

**Comment by pthowe** [comment]
Trying to compile JNativeHook,
Getting error when trying to compile

C:\Users\<path>\jnativehook-2.2.0>cmake -B target\build\libuiohook\ -S src\external\libuiohook\ -G "Visual Studio 16 2019" -A x64 -D CMAKE_INSTALL_PREFIX=target\ -D CMAKE_VERBOSE_MAKEFILE=ON -D BUILD_SHARED_LIBS=OFF
-- Selecting Windows SDK version 10.0.16299.0 to target Windows 10.0.19041.
-- The C compiler identification is MSVC 19.27.29111.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
CMake Error at CMakeLists.txt:47 (message):
  Could not determine target architecture.  Please set CMAKE_SYSTEM_PROCESSOR
  to the desired resource folder.


-- Configuring incomplete, errors occurred!
See also "C:/Users/<path>/jnativehook-2.2.0/target/build/libuiohook/CMakeFiles/CMakeOutput.log".

Any suggestions

**Comment by kwhat** [comment]
Sometimes cmake has a hard time figuring out what target you are building for.  Try something like this:

          cmake -B target\build\libuiohook\  -S src\external\libuiohook\ ^
            -G "Visual Studio 16 2019" -A x64 ^
            -D CMAKE_SYSTEM_PROCESSOR=x86_64 ^
            -D CMAKE_INSTALL_PREFIX=target\ ^
            -D CMAKE_VERBOSE_MAKEFILE=ON ^
            -D BUILD_SHARED_LIBS=OFF

_Edit: copy paste fail_  You need to add CMAKE_SYSTEM_PROCESSOR which is basically the output folder that is used in the resource folder.  You could also need `-D CMAKE_SYSTEM_NAME=windows` which is the OS name at the same location.

![image](https://user-images.githubusercontent.com/676164/136663777-7bb6292b-56ec-4bc2-89a7-81f15c0bac7b.png)

**Comment by kwhat** [comment]
I am kind of wondering what CMAKE_SYSTEM_PROCESSOR you are currently getting, can you patch [CMakeLists.txt:47](https://github.com/kwhat/jnativehook/blob/2.2/CMakeLists.txt#L47) from 

`message(FATAL_ERROR "Could not determine target architecture.  Please set CMAKE_SYSTEM_PROCESSOR to the desired resource folder.")` 

to 

`message(FATAL_ERROR "Could not determine target architecture. (${CMAKE_SYSTEM_PROCESSOR}) Please set CMAKE_SYSTEM_PROCESSOR to the desired resource folder.")`

and paste the output?

**Comment by pthowe** [comment]
I made the suggested change to CMakeLists.txt. 

C:\Users\<path>\jnativehook-2.2.0>cmake -B target\build\libuiohook\ -S src\external\libuiohook\ -G "Visual Studio 16 2019" -A x64 -D CMAKE_SYSTEM_PROCESSOR=AMD64 -D CMAKE_INSTALL_PREFIX=target\ -D CMAKE_VERBOSE_MAKEFILE=ON -D BUILD_SHARED_LIBS=OFF
-- Selecting Windows SDK version 10.0.16299.0 to target Windows 10.0.19041.
CMake Error at CMakeLists.txt:47 (message):
  Could not determine target architecture.  (AMD64) Please set
  CMAKE_SYSTEM_PROCESSOR to the desired resource folder.


-- Configuring incomplete, errors occurred!
See also "C:/Users/<path>/jnativehook-2.2.0/target/build/libuiohook/CMakeFiles/CMakeOutput.log".

C:\Users\<path>\jnativehook-2.2.0>

I tried assigning other values to CMAKE_SYSTEM_PROCESSOR nothing changed.
The compile creates a log I have attached.
[CMakeOutput-edited.log](https://github.com/kwhat/jnativehook/files/7316565/CMakeOutput-edited.log)

What files are the source for this operation? Is it possible I am missing files?

**Comment by kwhat** [comment]
> Could not determine target architecture. (AMD64) Please set CMAKE_SYSTEM_PROCESSOR to the desired resource folder.

Haha, its just a case sensitivity issue AMD64 != amd64.

**Comment by kwhat** [comment]
let me make some patches

**Comment by kwhat** [comment]
Alright, [this](https://github.com/kwhat/jnativehook/pull/368/files) change will probably fix the "Please set
CMAKE_SYSTEM_PROCESSOR to the desired resource folder" error.  I think that is the only error you are getting.

**Comment by kwhat** [comment]
Alright changes pushed to 2.2.  Go ahead pull the changes and let me know if there is still a problem.

**Comment by pthowe** [comment]
Still working on this issue. Your changes seemed to have helped.

C:\Users\<path>\GitHub\jnativehook>cmake -B target/build/libuiohook/ -S src/external/libuiohook/ \
-- Selecting Windows SDK version 10.0.16299.0 to target Windows 10.0.19041.
CMake Error at CMakeLists.txt:80 (find_package):
  By not providing "Finduiohook.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "uiohook", but
  CMake did not find one.

  Could not find a package configuration file provided by "uiohook" with any
  of the following names:

    uiohookConfig.cmake
    uiohook-config.cmake

  Add the installation prefix of "uiohook" to CMAKE_PREFIX_PATH or set
  "uiohook_DIR" to a directory containing one of the above files.  If
  "uiohook" provides a separate development package or SDK, be sure it has
  been installed.


-- Configuring incomplete, errors occurred!
See also "C:/Users/<path>/GitHub/jnativehook/target/build/libuiohook/CMakeFiles/CMakeOutput.log".

Where do I get uiohookConfig.cmake

[CMakeOutput-path.log](https://github.com/kwhat/jnativehook/files/7316930/CMakeOutput-path.log)
?

**Comment by kwhat** [comment]
It seems like find_package is not working correctly...  Did you checkout the libuiohook submodule?  Double check that you have src/external/libuiohook/CMakeLists.txt.

> Where do I get uiohookConfig.cmake

You need to compile uiohook first, and then install it to target\build\libuiohook\.  `-D CMAKE_PREFIX_PATH=target\ ^` is what should tell cmake where to find that file.


    md target\build\libuiohook\
    
    cmake -B target\build\libuiohook\  -S src\external\libuiohook\ ^
        -G "Visual Studio 16 2019" -A x64 ^
        -D CMAKE_INSTALL_PREFIX=target\ ^
        -D CMAKE_VERBOSE_MAKEFILE=ON ^
        -D BUILD_SHARED_LIBS=OFF
    
    cmake --build target\build\libuiohook\ ^
        --parallel 2 ^
        --config RelWithDebInfo ^
        --clean-first
    
    cmake --install target\build\libuiohook\ ^
        --config RelWithDebInfo

**Comment by pthowe** [comment]
Well I am getting closer I think.
I need to check out:
https://github.com/kwhat/jnativehook.git
https://github.com/kwhat/libuiohook.git
Compile libuiohook using: 

`C:\Users\<path>\GitHub\libuiohook>mkdir build

C:\Users\<path>\GitHub\libuiohook>cd build

C:\Users\<path>\GitHub\libuiohook\build> cmake -S .. -D BUILD_SHARED_LIBS=ON -D BUILD_DEMO=ON -DCMAKE_INSTALL_PREFIX=../dist
-- Building for: Visual Studio 16 2019
-- Selecting Windows SDK version 10.0.16299.0 to target Windows 10.0.19041.
-- The C compiler identification is MSVC 19.27.29111.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - not found
-- Found Threads: TRUE
-- Configuring done
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
ADVAPI32
    linked by target "uiohook" in directory C:/Users/Peter Howe/Documents/GitHub/libuiohook
C:/Users/Peter Howe/Documents/GitHub/libuiohook/ADVAPI32
   used as include directory in directory C:/Users/Peter Howe/Documents/GitHub/libuiohook

-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.

C:\Users\<path>\GitHub\libuiohook\build>`

The output from this will need to be inserted into the jnativehook/src/external/libuiohook

Is this the correct direction?

**Comment by pthowe** [comment]
Found compiled DLL,
Tried to add to JAR file. Have been unsuccessful.
Was able to update path to include DLL directory. This worked but would like to include DLL into JAR

Can this be accomplished.

**Comment by kwhat** [comment]
Dll's go in the resource folder, for example: src/main/resources/com/github/kwhat/jnativehook/lib/windows/x86_64/JNativeHook.dll
